text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,440 @@
+  principal.manager
+else
+  principal
+```
+
+Suppose we are validating at level 2, then this expression will have type `User(1)`, where we use subtyping to give expression `principal` (in the else branch) type `User(1)` rather than `User(2)`. Similarly, if we were defining a set of entities, all of them must be given the same level. For example, we could give `[ principal.manager, principal ]` the type `Set<User(1)>`. Note that for records, individual attributes may have entities with different levels. E.g., `{ a: principal, b: principal.manager }` would have type `{ a: User(2), b: User(1) }`.
+
+### Slicing 
+
+Once a schema has been validated against a non- infinite level, we can use the following procedure at runtime to compute the entity slice: 
+
+```
+/// Takes:
+/// request : The request object
+/// level : The level the schema was validated at 
+/// get-entity : A function for looking up entity data given an EUID 
+function slice(request, level, get-entity) { 
+    // This will be our entity slice
+   let mut es : Map<EUID, EntityData> = {};
+
+   let mut worklist : Set<EUID> = {principal, action, resource, all root euids in context};
+   for euid in worklist {
+     es.insert(euid, get-entity(euid));
+   }
+   
+   let current-level = 0;
+   
+   while current-level <= level || worklist.empty? {
+        // `t` will become our new worklist
+        let t : Set<EUID> = {};
+        // Clear our worklist
+        for euid in worklist {
+            if !es.contains(euid) {
+                es.insert(euid, get-entity(euid));
+            }
+            ; Set of all euids mentioned in this entities data
+            let new-euids = filter(not-in-es?, all-euids(es.get(euid));
+            t = t.union(new-euids);
+        }
+        // set t to be the new worklist
+        s = t;
+        // increment level
+        level = level + 1;
+   }
+
+    return es;
+}
+
+```
+
+### Soundness
+
+Leveled types permit us to generalize the soundness statement we use today. Currently, soundness assumes that expressions like `principal.manager` will never error, i.e., that the proper entity information is available. Now we assume dereferences are only available up to a certain level. In particular, we define a well-formedness judgment μ⊢v:τwhere μ is an entity store, v is a value, and τ is a possibly leveled type. This judgment is standard for values having non-entity type, e.g.,
+
+```
+mu |- i : Long
+
+mu |- s : String
+
+mu |- v1 : T1 ... mu |- vn : Tn 
+-------------------------------------------------------
+mu |- { f1: v1, ..., fn: vn } : { f1: T1, ..., fn: Tn }
+```
+
+There are two rules for entity literals of type `E::s`. The first says level `0` is always allowed. The second says a literal can have a level `n` only if `n` dereferences are possible starting from `E::s`:
+
+```
+mu |- E::s : E(0)
+
+mu(E::s) = (v,h) where v = { f1: v1, ..., fn: vn } and h is an ancestor set
+mu |- v1 : T1 ... mu |- vn : Tn
+where level(T1) >= n-1
```

Comment:
Also need `level(Tn) >= n-1` right?. Also, I think you might want to use a different index for the attributes to avoid confusion with the level `n`